# (연습) 변수, 값, 표현식

**참고 사항**

먼저 
[변수, 값, 표현식](https://codingalzi.github.io/pybook/variables_expressions.html)의 내용과
[(필수 예제) 변수, 값, 표현식](https://colab.research.google.com/github/codingalzi/pybook/blob/master/examples/examples-variables_expressions.ipynb)의 예제들을 학습하세요.

**주의 사항**

* 기존에 작성된 코드셀/텍스트셀은 수정하지 않는다.
* 필요한 경우 코드셀 또는 텍스트셀을 추가할 수 있다.
* 실습 파일을 구글 드라이브에서 링크로 누구가 공유할 수 있도록 설정하세요.

## 문제 1

아래 문장들을 적절한 연산자를 이용한 표현식으로 작성한 다음 실행 결과를 확인하라.

**표현식 1:** 정수 13에서 정수 19 빼기

**답**

뺄셈 연산자를 활용한다.

In [1]:
13 - 19

-6

**표현식 2:** 정수 81을 정수 3으로 나누기

**답**

나눗셈 연산자를 활용한다. 결과는 항상 부동소수점이다.

In [5]:
81/3

27.0

**표현식 3:** 정수 100을 정수 7로 나눈 나머지

**답**

나머지 연산자를 활용한다. 결과는 항상 정수다.

In [6]:
100 % 7

2

**표현식 4:** 정수 7.2의 세제곱

**답**

거듭제곱 연산자를 활용한다.
소수점 이하 맨끝자리에서 보이는 5는 부동소수점 연산의 한계를 보여준다.
원래 0이어야 한다.

In [7]:
7.2**3

373.24800000000005

## 문제 2

오전 6시 52분에 집을 떠나서, 가볍게 (km당 8분 15초 소요) 1km를 뛰고, 
이어서 좀 빠르게 (km당 7분 12초 소요) 3km를 뛴 후, 다시 가볍게 1km를 마저 달려 집으로 돌아오면 
집에 도착하는 시간을 계산하여 아래처럼 출력하는 코드를 작성하라.

    집에 도착 시간: 7 시 30 분 6 초

힌트: 정수 나눗셈의 몫과 나머지 연산자 활용

**답**

시, 분, 초를 동시에 다뤄야 하기에 계산을 초 단위로 통일시킨다.
그렇지 않으면 시간의 연산이 복잡해진다.
먼저 출발시간인 6시 52분이 밤 0시를 기준으로 몇 초 지난 시간인지
계산해서 `start` 변수에 할당한다.

In [66]:
start = (6 * 60 + 52) * 60
start

24720

아래 두 변수는 1 킬로미터를 가볍게 달릴 때와 좀 빠르게 달릴 때 걸리는 시간(초)을 가리킨다.

In [67]:
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

그러면 달리는 시간을 초로 계산하면 2286초로 나온다.

In [68]:
running_time = (1 + 1) * slow_running + 3 * fast_running
running_time

2286

도착시간은 밤 0시를 기준으로 출발시간에 달린 시간을 더하면 된다.

In [69]:
arrival = start + running_time
arrival

27006

도착시간을 시, 분, 초로 변환하기 위해 1분은 60초, 1시간은 60분이라는 정보를
이용하여 도착시간에 대해 60으로 나눴을 때의 몫과 나머지 계산을 두 번 연속 적용한다.

- 도착시간의 초 계산: 60으로 나눴을 때의 나머지

In [70]:
arrival_seconds = arrival % 60
arrival_seconds

6

- 도착시간의 분 계산: 60으로 나눴을 때의 몫

In [71]:
arrival_minutes = arrival // 60
arrival_minutes

450

450분은 60분 보다 크기에 시와 분으로 다시 변환해야 한다.
한 시간이 60분이기에 역시 60으로 나눈 몫과 나머지를 계산한다.

- 도착시간의 시 계산: 60으로 나눴을 때의 몫

In [72]:
arrival_hours = arrival_minutes // 60
arrival_hours

7

- 도착시간의 분 계산: 60으로 나눴을 때의 나머지

In [73]:
arrival_minutes = arrival_minutes % 60
arrival_minutes

30

따라서 도착시간은 7시 30분 6초이다. 

In [74]:
print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


정리하면 다음과 같다.

In [75]:
start = (6 * 60 + 52) * 60
slow_running = 8 * 60 + 15 # 가볍게 달릴 때 킬로미터 당 걸리는 시간(초)
fast_running = 7 * 60 + 12 # 좀 빠르게 달릴 때 킬로미터 당 걸리는 시간(초)

running_time = (1 + 1) * slow_running + 3 * fast_running
arrival = start + running_time

arrival_seconds = arrival % 60
arrival_minutes = arrival // 60
arrival_hours = arrival_minutes // 60
arrival_minutes = arrival_minutes % 60

print("집에 도착 시간:", arrival_hours, "시", arrival_minutes, "분", arrival_seconds, "초")

집에 도착 시간: 7 시 30 분 6 초


## 문제 3

두 개의 부동소수점`x`와 `y`를 입력받아 두 수의 
조화 평균을 아래 형식으로 출력하는 코드를 작성하라.

    2.4 와 5.7 의 조화평균: 3.38

단, `x`와 `y`의 조화평균은 다음과 같이 계산하며, 소수점 이하 셋째 자리에서 반올림한다.

    2 / (1/x + 1/y)

힌트: `float()` 함수와 `round()` 함수 활용

**답**

사용자가 정수가 아닌 유한소수를 입력할 것을 예상해야 하기에
`input()` 함수의 결과에 바로 `float()` 함수를 적용한다.
그 다음에 평균값을 계산하여 지정된 방식으로 출력한다.
단, `round()` 함수를 이용하여 소수점 이하의 자릿수를 2로 지정한다.

In [16]:
x = float(input("첫째 유한소수를 입력하세요: "))
y = float(input("둘째 유한소수를 입력하세요: "))

# 평균값
mean_harmony = 2 / (1/x + 1/y)

print(x, "와", y, "의 조화평균:", round(mean_harmony, 2))

첫째 유한소수를 입력하세요: 2.4
둘째 유한소수를 입력하세요: 5.7
2.4 와 5.7 의 조화평균: 3.38


## 문제 4

입력받은 두 유한소수의 나눗셈을 아래 지정된 형식으로 출력하는 코드를 작성하라.
입력된 두 개의 유한소수는 각각 변수 `a`와 `b`에 저장한다.
예를 들어 3.3과 5.8이 입력되면 아래처럼 화면에 출력한다.
또한 계산 결과는 소수점 이하 셋째 자리에서 반올림한다.

    3.3 / 5.8 = 0.57

**답**

유한소수 두 개를 각각 입력받을 수 있도록 `input()` 함수를 `float()` 함수와 함께 두 번 사용한다
첫째 입력값은 정수로 변환해서 `a`에 할당, 둘째 입력값은 `b`에 할당한다.
지정된 형식으로의 출력은 `print()` 함수를 이용한다.
`print()` 함수에 쉼표를 이용하여 여러 개의 인자를 지정하면
각각의 인자가 공백으로 구분되어 한 줄에 출력되는 성질을 이용한다.

In [20]:
a = float(input("첫째 정수를 입력하세요: "))
b = float(input("둘째 정수를 입력하세요: "))

print(a, "/", b, "=", round(a / b, 2))

첫째 정수를 입력하세요: 3.3
둘째 정수를 입력하세요: 5.8
3.3 / 5.8 = 0.57


## 문제 5

사용자로부터 입력받은 후 입력된 정수에 대한 3의 배수 여부를 판정하는 코드를 작성하라.
단, 3의 배수인 경우와 아닌 경우 각각 다르게 아래 형식으로 출력되도록 한다.

    24 은(는) 3의 배수다.
    
또는

    25 은(는) 3으로 나누면 나머지가 1 이다.


힌트: `input()` 함수, `int()` 함수, 몫과 나머지 연산자 활용. `if ... else ...` 조건문 사용.

**답**

3의 배수 여부는 3으로 나눈 나머지에 의존한다.
나머지가 0일 때만 3의 배수다.
나머지는 나머지 연산자를 이용하여 확인한다.

In [9]:
user_input = int(input("정수 하나를 입력하세요: "))

if user_input % 3 == 0:
    print(user_input, "은(는)", "3의 배수다.")
else:
    residual = user_input % 3
    print(user_input, "은(는)", "3으로 나누면 나머지가", residual, "이다.")

정수 하나를 입력하세요: 25
25 은(는) 3으로 나누면 나머지가 1 이다.


## 문제 6

대기실에 `N`개의 의자가 일렬로 놓여 있다.
어느 두 사람도 이웃하여 앉지 않을 때, 앉을 수 있는 최대 인원수와
최소 인원수를 출력하는 코드를 구하라.
입력은 100이하의 자연수이고, 최대와 최소는 공백으로 구분하여 출력한다.
단, 자리를 채운 후에는 더 이상 어느 누구도 자리에 앉을 수 없어야 한다.
예를 들어 의자가 8개인 경우 아래 방식으로 앉는 것은 허용되지 않는다.
아래 표에서 `O`은 채워진 자리를, `X`는 빈자리를 가리킨다.

| 불가능 방식 | 이유 |
| :---: | :--- |
| `XXOXOXOX` | 왼쪽 끝에 한 명 더 앉을 수 있음 |
| `XOXXOXXX` | 오른쪽 끝에 한 명 더 앉을 수 있음|

의자가 8 개일 때 의자에 앉을 수 있는 최소 인원과 최대 인원은 다음과 같다.

| 인원 | 앉는 방식 |
| :---: | :---: |
| 최소 3명 | `XOXXOXXO` |
| 최대 4명 | `OXOXOXOX` |

힌트: 2배수와 3배수 여부에 따른 `if ... else ...` 조건문 활용

**답**

`N`개의 의자가 있을 때 예를 들어 맨 왼쪽 의자에 앉은 다음에 한 자리 건널 때마다 앉으면 최대 인원이 앉게 된다.
즉 두 의자당 하나에 앉으면 된다. 만약 `N`이 홀수이면 맨 오른쪽 자리에 한 명이 더 앉을 수 있다.
따라서 `N`이 2배수인지 여부가 중요하다.

- `N`이 2의 배수일 때: 최대 인원은 `N//2`
- `N`이 2의 배수일 때: 최대 인원은 `N//2 + 1`

반면에 최소 인원은 다음과 같이 계산할 수 있다. 
먼저 의자가 1개 또는 2개이면 최대 1명 앉을 수 있다.
이제부터 의자가 3개 이상이라고 가정한다.
먼저 `N = 3`이라고 하자. 그러면 아래 방식으로 1명만 앉을 수 있다.

    XOX

`N > 3`인 경우는 3배수 여부로 결정된다.

- `N`이 3의 배수인 경우: `XOX` 방식을 `N//3` 번 반복한다. 예를 들어 `N = 6`이면 아래 방식으로 최소 인원이 앉는다.

        XOXXOX
        
- `N`이 3의 배수가 아닌 경우: `XOX` 방식을 `N//3` 번 반복한 다음 나머지 하나 또는 두 자리 중에 한 자리에 앉는다.
    예를 들어 `N=8`이면 아래 방식으로 최소 인원이 앉는다.

        XOXXOXXO

위 설명을 정리하여 코드로 작성하면 다음과 같다.

In [1]:
N = int(input("정수를 입력하세요: "))

# 최대 인원
if (N%2 == 0) :
    max = N//2
else :
    max = (N//2) + 1

# 최소 인원
if (N%3 == 0) :
    min = N//3
else :
    min = (N//3) + 1

print("최소:", min, "명")
print("최대:", max, "명")

정수를 입력하세요: 8
최소: 3 명
최대: 4 명


## 문제 7

다음 모양을 출력하는 코드를 작성하라.
두 개의 별 기호 사이에 공백(space)가 사용됨에 주의한다.
공백도 하나의 기호로 취급된다.
단, `print()` 함수와 `while` 반복문은 한 번씩만 활용한다.

```
        * 
      * * 
    * * * 
  * * * * 
* * * * * 
  * * * * 
    * * * 
      * * 
        * 
```

**답**

별 기호와 공백으로 구성된 문자열, 즉 `"* "`이 차례대로 1, 2, 3, 4, 5 번씩,
이후에는 4, 3, 2, 1 번씩 반복된다.
`while` 반복문:  총 9번 출력하며 5번째 출력을 중심으로 처음 4번과 나중 4번이 대칭을 이룬다.

`num`이 1부터 9까지 변할 때 두 개의 공백으로 구성된 문자열 `"  "`와
별과 공백으로 구성된 문자열 `'* '`이 반복되는 횟수를 각각 `spaces` 변수와 `stars` 변수에 할당한 후에
이를 이용하여 출력문을 실행한다.

- `num <= 5` 일 때
    - 공백: 총 `5 - num` 개
    - 별: 총 `num` 개
- `num > 5` 일 때
    - 공백: 총 `num - 5` 개
    - 별: 총 `10 - num` 개

In [2]:
num = 1

while num < 10:
    if num <= 5:
        spaces = 5 - num
        stars = num
    else:
        spaces = num - 5
        stars = 10 - num
        
    print("  " * spaces + "* " * stars)
    
    num += 1

        * 
      * * 
    * * * 
  * * * * 
* * * * * 
  * * * * 
    * * * 
      * * 
        * 


## 문제 8

위 문제를 `for` 반복문을 한 번만 활용하여 해결하라.
단, `print()` 함수와 `for` 반복문은 한 번씩만 활용한다.

```
        * 
      * * 
    * * * 
  * * * * 
* * * * * 
  * * * * 
    * * * 
      * * 
        * 
```

힌트: `range()` 함수 활용

**답**

`for` 반복문을 `num` 변수가 1부터 9까지에 대해 실행하도록 한다.
이를 위해 `range(1, 10)` 함수 표현식을 이용한다.

In [1]:
for num in range(1, 10):
    if num <= 5:
        spaces = 5 - num
        stars = num
    else:
        spaces = num - 5
        stars = 10 - num
        
    print("  " * spaces + "* " * stars)

        * 
      * * 
    * * * 
  * * * * 
* * * * * 
  * * * * 
    * * * 
      * * 
        * 
